pricegraph/src/orderbook.rs

```diff
@@ -126,22 +126,24 @@ impl Orderbook {
     }
 
     /// Reduces the orderbook by matching all overlapping ring trades.
-    pub fn reduce_overlapping_orders(mut self) -> ReducedOrderbook {
-        Subgraphs::new(self.projection.node_indices()).for_each(|token| loop {
+    pub fn reduce_overlapping_orders(mut self) -> Result<ReducedOrderbook, OrderbookError> {
+        let result = Subgraphs::new(self.projection.node_indices()).for_each_until(|token| loop {
             let cycle = match shortest_path(&self.projection, token, None) {
-                Ok(shortest_path_graph) => break shortest_path_graph.connected_nodes(),
+                Ok(shortest_path_graph) => {
+                    break ControlFlow::Continue(shortest_path_graph.connected_nodes())
+                }
                 Err(cycle) => cycle,
             };
-            self.fill_path(&cycle).unwrap_or_else(|| {
-                panic!(
-                    "failed to fill path along detected negative cycle {}",
-                    format_path(&cycle),
-                )
```

Comment:
This should still be a panic as it is indicates the logic in the code is wrong.

Reply:
The root problem is that find_path_flow can create an invalid flow in which case we return the error. fill_path calls find_path_flow so we cannot unwrap here.

Reply:
Hmm, maybe the errors need to be differentiated so that we can panic if its a negative cycle error but break if its a floating point error.